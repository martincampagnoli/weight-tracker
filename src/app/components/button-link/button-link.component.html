<a class="card w-100 pb-3 pt-3" rel="noopener" [routerLink]="route()">
  <mat-icon>{{ icon() }}</mat-icon>
  <span class="ps-1">{{ label() }}</span>
  <mat-icon>chevron_right</mat-icon>
</a>
