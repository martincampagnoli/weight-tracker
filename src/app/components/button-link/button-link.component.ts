import { ChangeDetectionStrategy, Component, input } from '@angular/core';
import { MatIconModule } from '@angular/material/icon';
import { RouterLink } from '@angular/router';

/**
 * Component representing a button-styled link with label, route, and icon.
 */
@Component({
  selector: 'app-button-link',
  templateUrl: './button-link.component.html',
  styleUrls: ['./button-link.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
  standalone: true,
  imports: [MatIconModule, RouterLink],
})
export class ButtonLinkComponent {
  /**
   * The label text for the button-styled link.
   */
  readonly label = input.required<string>();

  /**
   * The route URL for navigation when the button-styled link is clicked.
   */
  readonly route = input.required<string>();

  /**
   * The name or class of the icon to be displayed within the button-styled link.
   */
  readonly icon = input.required<string>();

  /**
   * Constructs the ButtonLinkComponent.
   */
  constructor() {}
}
