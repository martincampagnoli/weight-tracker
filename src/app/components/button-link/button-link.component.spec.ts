import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ButtonLinkComponent } from './button-link.component';
import { MatIconModule } from '@angular/material/icon';
import { RouterTestingModule } from '@angular/router/testing';

describe('ButtonLinkComponent', () => {
  let component: ButtonLinkComponent;
  let fixture: ComponentFixture<ButtonLinkComponent>;
  let element: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [ButtonLinkComponent, MatIconModule, RouterTestingModule],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(ButtonLinkComponent);
    component = fixture.componentInstance;
    element = fixture.nativeElement;

    fixture.componentRef.setInput('label', 'Test Label');
    fixture.componentRef.setInput('route', '/test-route');
    fixture.componentRef.setInput('icon', 'fa fa-icon');

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
  it('should display the label', () => {
    const labelElement = element.querySelector('span');
    expect(labelElement?.innerHTML).toContain('Test Label');
  });

  it('should have a route', () => {
    const linkElement = element.querySelector('a');
    expect(linkElement?.getAttribute('href')).toBe('/test-route');
  });
  it('should display the icon', () => {
    const iconElement = element.querySelector('mat-icon');
    expect(iconElement).toBeTruthy();
  });

  it('should have the correct routerLink attribute', () => {
    const linkElement = element.querySelector('a');
    expect(linkElement).toBeTruthy();
    expect(component).toBeTruthy();
  });

  it('should match snapshot with default props', () => {
    expect(element).toMatchSnapshot();
  });

  it('should match snapshot with different icon', () => {
    const newFixture = TestBed.createComponent(ButtonLinkComponent);
    newFixture.componentRef.setInput('label', 'Settings');
    newFixture.componentRef.setInput('route', '/settings');
    newFixture.componentRef.setInput('icon', 'settings');
    newFixture.detectChanges();
    expect(newFixture.nativeElement).toMatchSnapshot();
  });

  it('should match snapshot with long label', () => {
    const newFixture = TestBed.createComponent(ButtonLinkComponent);
    newFixture.componentRef.setInput(
      'label',
      'This is a very long button label'
    );
    newFixture.componentRef.setInput('route', '/long-route-name');
    newFixture.componentRef.setInput('icon', 'info');
    newFixture.detectChanges();
    expect(newFixture.nativeElement).toMatchSnapshot();
  });

  it('should match snapshot with empty-like values', () => {
    const newFixture = TestBed.createComponent(ButtonLinkComponent);
    newFixture.componentRef.setInput('label', '');
    newFixture.componentRef.setInput('route', '/');
    newFixture.componentRef.setInput('icon', '');
    newFixture.detectChanges();
    expect(newFixture.nativeElement).toMatchSnapshot();
  });
});
