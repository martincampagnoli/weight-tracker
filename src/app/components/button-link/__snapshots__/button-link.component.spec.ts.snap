// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonLinkComponent should match snapshot with default props 1`] = `
<div
  id="root5"
>
  <a
    class="card w-100 pb-3 pt-3"
    href="/test-route"
    rel="noopener"
  >
    <mat-icon
      aria-hidden="true"
      class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
      data-mat-icon-type="font"
      role="img"
    >
      fa fa-icon
    </mat-icon>
    <span
      class="ps-1"
    >
      Test Label
    </span>
    <mat-icon
      aria-hidden="true"
      class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
      data-mat-icon-type="font"
      role="img"
    >
      chevron_right
    </mat-icon>
  </a>
</div>
`;

exports[`ButtonLinkComponent should match snapshot with different icon 1`] = `
<div
  id="root7"
>
  <a
    class="card w-100 pb-3 pt-3"
    href="/settings"
    rel="noopener"
  >
    <mat-icon
      aria-hidden="true"
      class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
      data-mat-icon-type="font"
      role="img"
    >
      settings
    </mat-icon>
    <span
      class="ps-1"
    >
      Settings
    </span>
    <mat-icon
      aria-hidden="true"
      class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
      data-mat-icon-type="font"
      role="img"
    >
      chevron_right
    </mat-icon>
  </a>
</div>
`;

exports[`ButtonLinkComponent should match snapshot with empty-like values 1`] = `
<div
  id="root11"
>
  <a
    class="card w-100 pb-3 pt-3"
    href="/"
    rel="noopener"
  >
    <mat-icon
      aria-hidden="true"
      class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
      data-mat-icon-type="font"
      role="img"
    >
      
    </mat-icon>
    <span
      class="ps-1"
    >
      
    </span>
    <mat-icon
      aria-hidden="true"
      class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
      data-mat-icon-type="font"
      role="img"
    >
      chevron_right
    </mat-icon>
  </a>
</div>
`;

exports[`ButtonLinkComponent should match snapshot with long label 1`] = `
<div
  id="root9"
>
  <a
    class="card w-100 pb-3 pt-3"
    href="/long-route-name"
    rel="noopener"
  >
    <mat-icon
      aria-hidden="true"
      class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
      data-mat-icon-type="font"
      role="img"
    >
      info
    </mat-icon>
    <span
      class="ps-1"
    >
      This is a very long button label
    </span>
    <mat-icon
      aria-hidden="true"
      class="mat-icon notranslate material-icons mat-ligature-font mat-icon-no-color"
      data-mat-icon-type="font"
      role="img"
    >
      chevron_right
    </mat-icon>
  </a>
</div>
`;
